miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/api/ApiClient.kt

```diff
@@ -63,52 +64,79 @@ internal class ApiClient @VisibleForTesting constructor(
-                }
+    @Suppress("TooGenericExceptionCaught", "ThrowsCount")
+    suspend fun <T> executeRequest(call: Call<T>): T = try {
+        val response = call.execute()
+        when {
+            response.isSuccessful -> {
+                // Body shouldn't be null if request was successful
+                response.body() ?: throw sdkExceptionForInternalServerError()
             }
-        } catch (error: Exception) { // when response is not Type T or malformed JSON is received
-            throw MiniAppSdkException(error)
+            else -> throw exceptionForHttpError<T>(response)
         }
+    } catch (error: Exception) { // when response is not Type T or malformed JSON is received
+        throw MiniAppSdkException(error)
     }
 
-    private fun sdkExceptionFromHttpException(
-        response: Response<in Nothing>,
-        error: ResponseBody
-    ): MiniAppSdkException {
-        return MiniAppSdkException(
-            MiniAppHttpException(
-                response = response,
-                errorMessage = errorConvertor.convert(error)?.message
-                    ?: "No error message provided by server."
-            ).errorMessage
-        )
+    @Throws(MiniAppSdkException::class)
+    @Suppress("MagicNumber", "ThrowsCount")
+    private fun <T> exceptionForHttpError(response: Response<T>): MiniAppSdkException {
+        // Error body shouldn't be null if request wasn't successful
+        val errorData = response.errorBody() ?: throw sdkExceptionForInternalServerError()
+        when (response.code()) {
+            401, 403 -> throw MiniAppSdkException(
+                convertApiCatalogErrorToMsg(
+                    response, errorData, createErrorConvertor(retrofit)
+                )
+            )
+            else -> throw MiniAppSdkException(
+                convertStandardHttpErrorToMsg(
+                    response, errorData, createErrorConvertor(retrofit)
+                )
+            )
+        }
     }
+
+    private fun convertApiCatalogErrorToMsg(
+        response: Response<in Nothing>,
+        error: ResponseBody,
+        converter: Converter<ResponseBody, ApiCatalogErrorResponse>
+    ) = errorMsgFromHttpException(response, converter.convert(error)?.message)
+
+    private fun convertStandardHttpErrorToMsg(
+        response: Response<in Nothing>,
+        error: ResponseBody,
+        converter: Converter<ResponseBody, HttpErrorResponse>
+    ) = errorMsgFromHttpException(response, converter.convert(error)?.message)
+
+    private fun errorMsgFromHttpException(
+        response: Response<in Nothing>,
+        error: String?
+    ) = MiniAppHttpException(
+        response = response,
+        errorMessage = error ?: "No error message provided by server."
+    ).message()
 }
 
-internal data class ErrorResponse(
+internal data class HttpErrorResponse(
     val code: Int,
+    override val message: String
+) : ErrorResponse
+
+internal data class ApiCatalogErrorResponse(
```

Comment:
I think this name is exposing details about the API that we don't need to know as the SDK. How about `AuthErrorResponse` instead?

Reply:
On iOS we called it UnauthorizedData, as opposed to ErrorData we already had in place

Reply:
The choice I made was by considering future reference and readability. I would definitely forget why there are two types from the backend. The SDK is managing the deviation of response types from the backend. So, it is aware indeed.

I had two options, mark that domain-effect into the names of behavior or record it as a comment. I choose the former.

Unless, there is a strong opinion on renaming it, I would like to keep it the same.

Reply:
I think something that refers to it being related to Authorization is more explanatory. It's not the SDK's responsibility to document why certain choices were made by the API side. We just have to document that there is a different response for different response codes.

Also, keep in mind that this is a public SDK, so outside developers will have no idea what API Catalog is.

Reply:
Thanks @corycaywood, I think that's reasonable. 👍